src/finn/custom_op/fpgadataflow/thresholding_binary_search.py

```diff
@@ -0,0 +1,594 @@
+        assert np.equal(
+            np.mod(orig_thres_matrix, 1), 0
+        ).all(), "Need int threshold tensor"
+        ret = orig_thres_matrix
+        # ensure channels = mh , duplicating if necessary
+        if ret.shape[0] == 1:
+            ret = np.tile(ret, (mh, 1))
+        assert (
+            ret.shape[0] == mh
+        ), "Channels of threshold matrix are not as expected (mh)"
+        # distribute rows between PEs
+        ret = interleave_matrix_outer_dim_from_partitions(ret, pe)
+        assert (
+            ret.shape[0] == pe
+        ), """First dimension after distribution of the
+        rows between PEs is not as expected (pe)"""
+        assert (
+            ret.shape[1] == tmem
+        ), """Second dimension after distribution of the
+        rows between PEs is not as expected (tmem)"""
+        assert (
+            ret.shape[2] == n_thres_steps
+        ), """Third dimension after distribution of the
+        rows between PEs is not as expected (n_thres_steps)"""
+        return ret.reshape(1, pe, tmem, n_thres_steps)
+
+    def prepare_codegen_rtl_values(self):
+        """All dictionary values produced in this function are to replace
+        their key value(s) in the RTL template files"""
+        code_gen_dict = {}
+
+        # Identify the module names
+        code_gen_dict["$MODULE_NAME$"] = [self.get_verilog_top_module_name()]
+        code_gen_dict["$MODULE_NAME_AXI$"] = [
+            self.get_verilog_top_module_name() + "_axi"
+        ]
+        code_gen_dict["$MODULE_NAME_AXI_WRAPPER$"] = [
+            self.get_verilog_top_module_name() + "_axi_wrapper"
+        ]
+        # Set the top module name - AXI wrapper
+        code_gen_dict["$TOP_MODULE$"] = code_gen_dict["$MODULE_NAME_AXI_WRAPPER$"]
+
+        # Identify the module variables
+        output_data_type = self.get_nodeattr("outputDataType")  # output precision
+        input_data_type = self.get_nodeattr(
+            "inputDataType"
+        )  # input/threshold precision
+        num_channels = self.get_nodeattr("NumChannels")  # number of channels
+        bias = self.get_nodeattr("activation_bias")  # activation bias value
+
+        code_gen_dict["$N$"] = [
+            str(DataType[output_data_type].bitwidth())
+        ]  # output precision - convert bitwidth to string
+        code_gen_dict["$M$"] = [
+            str(DataType[input_data_type].bitwidth())
+        ]  # input/threshold precision - convert bitwidth to string
+        code_gen_dict["$C$"] = [str(num_channels)]  # number of channels
+        code_gen_dict["$BIAS$"] = [str(bias)]  # activation bias value
+
+        # Is the input datatype signed or unsigned?
+        # The thresholding core needs to know this when comparing weights to inputs
+        if self.get_input_datatype().signed():
+            code_gen_dict["$SIGN$"] = ["signed"]
+        else:
+            code_gen_dict["$SIGN$"] = ["unsigned"]
+
+        return code_gen_dict
+
+    def get_rtl_file_list(self):
```

Comment:
Please add docstrings for functions that are not in base class.

Reply:
Added - see https://github.com/Xilinx/finn/pull/687/commits/0d99b6c8ed358b2feea41cc8af242d40b30c8d97